### ISSUE TITLE
[consteval] clang++ sometimes reports false-positive errors for static consteval member functions

### ISSUE THREAD

**Issue by ldalessa** [llvm/llvm-project#56246]
When I have a CRTP that uses it's client's function, if that client's function happens to be `static consteval`, `clang++` reports an error.

```c++
template <class T>
struct base
{
    constexpr auto foobar() const -> int {
        return static_cast<T const*>(this)->bar(); // error: call to consteval function 'foo::bar' is not a constant expression
    }
};

struct foo : base<foo> {
    static consteval auto bar() -> int {
        return 0;
    }
};

foo x;
int i = x.foobar();
```

live: https://godbolt.org/z/KG5oT6GTj

**Comment by usx95** [comment]
For the example

**Comment by usx95** [comment]
This does not work even for simpler cases of `this` and `consteval` interactions.

```
struct S {
    static consteval int hello() { return 0;}

    int foo() const { 
        return this->hello(); 
        // ^ error: call to consteval function 'S::hello' is not a constant expression
        // ^ note: use of 'this' pointer is only allowed within the evaluation of a call to a 'constexpr' member function
    }
    constexpr int bar() const {
        return this->hello(); // same error and note as above.
    }
    consteval int baz() const {
        // works fine.
        return this->hello();
    }
};
```
Clang errors and diverges from all the rest of the compilers. https://godbolt.org/z/e9zo1E6e3

**Comment by usx95** [comment]
https://github.com/llvm/llvm-project/commit/9fcca8b470fb92466ec4cada9fbeacc90096116d#diff-f10defc3f20fb095bf22b3a79bead200d494bde9d503e283067a57aff483936c

I think this patch fixed a [crash](https://github.com/llvm/llvm-project/issues/47579) but clang still produces an error. 

>>>
```cpp
struct A {
  consteval int a() const { return 1; }
  void c() {
    a() + d(); // expected-error {{call to consteval function 'PR48235::A::a' is not a constant expression}} \
               // expected-note {{use of 'this' pointer is only allowed within the evaluation of a call to a 'constexpr' member function}}
  }
```
My reading of https://eel.is/c++draft/dcl.constexpr is that **consteval members** are valid and using them in any context (constant or unevaluated) show also be legal.

Would be interested in knowing thoughts of @AaronBallman

**Comment by AaronBallman** [comment]
> [9fcca8b#diff-f10defc3f20fb095bf22b3a79bead200d494bde9d503e283067a57aff483936c](https://github.com/llvm/llvm-project/commit/9fcca8b470fb92466ec4cada9fbeacc90096116d#diff-f10defc3f20fb095bf22b3a79bead200d494bde9d503e283067a57aff483936c)
> 
> I think this patch fixed a [crash](https://github.com/llvm/llvm-project/issues/47579) but clang still produces an error.
> 
> > > >
> 
> ```c++
> struct A {
>   consteval int a() const { return 1; }
>   void c() {
>     a() + d(); // expected-error {{call to consteval function 'PR48235::A::a' is not a constant expression}} \
>                // expected-note {{use of 'this' pointer is only allowed within the evaluation of a call to a 'constexpr' member function}}
>   }
> ```
> 
> My reading of https://eel.is/c++draft/dcl.constexpr is that **consteval members** are valid and using them in any context (constant or unevaluated) show also be legal.
> 
> Would be interested in knowing thoughts of @AaronBallman

I believe http://eel.is/c++draft/expr.prim.id.general#4 is the cause of that diagnostic. https://godbolt.org/z/PP51G3WWW sure suggests everyone implements this the same way, too.

**Comment by usx95** [comment]
Interesting. The above example looks right now.

Even the clang error for below is justified. (Other compilers diverge here: https://godbolt.org/z/bGbvh59oE)
```cpp
struct A {
    static consteval int a() { return 1; }
    int c() {
        return this->a();  // error
    }
};
```

Coming back to 
```cpp
struct S {
    static consteval int hello() { return 1;}

    int foo() const { 
        return this->hello(); 
        // ^ error: call to consteval function 'S::hello' is not a constant expression
        // ^ note: use of 'this' pointer is only allowed within the evaluation of a call to a 'constexpr' member function
    }
    constexpr int bar() const {
        return this->hello(); // same error and note as above.
    }
    consteval int baz() const {
        // works fine.
        return this->hello();
    }
} s;
static_assert(s.baz() == 1);

```
Then I guess clang is almost right but it should consider `this->hello()`  in `bar()` to be a **constant expression** according to http://eel.is/c++draft/dcl.constexpr#5.1 and http://eel.is/c++draft/expr.const#5.1

**Comment by AaronBallman** [comment]
I think you're correct because of http://eel.is/c++draft/expr.const#5.1.2 specifically. While the `this` isn't necessary for the member access expression (because the expression denotes a static function), that `this` is still evaluated per https://eel.is/c++draft/expr.ref#1.sentence-2 (see the associated footnote as well), so it is being used as the postfix-expression and so it should be fine as a core constant expression.

So I think the diagnostic in `foo()` is correct, but the one in `bar()` is incorrect, and accepting the code in `baz()` is correct. Do you come to the same conclusion?

**Comment by usx95** [comment]
I come to the conclusion that 
`this->static_consteval_member()` is a **core constant expression** and thus the invocation inside `foo()` is also valid and clang diagnostic is incorrect. \
http://eel.is/c++draft/dcl.constexpr#5.1 : `an invocation of a constexpr function can appear in a constant expression [expr.const]`

I agree with the rest of the conclusion (`bar()` and `baz()`).

**Comment by usx95** [comment]
Although I suspect I might be wrong again due to http://eel.is/c++draft/expr.prim.id.general#4. I am mostly confused about the definition of `id-expression`. Basically:
```cpp
consteval int func() {return 0;}
struct S {
    static consteval int hello() { return 1;}
    int foo() const { 
        return this->hello(); // Why this should be invalid
        // return func() ; // but this is valid. Both are core constant expressions.
    }
}
```
Is `func()` not an id-expression and http://eel.is/c++draft/expr.prim.id.general#4 does not apply to it ?

**Comment by AaronBallman** [comment]
`func` is an `id-expression` (it's the `primary-expression` component of the `postfix-expression` and the `()` denotes a function call expression http://eel.is/c++draft/expr.compound#nt:postfix-expression). But `id-expression` is also used in a `postfix-expression` denoting a member access in the same paragraph:
```
postfix-expression . template_opt id-expression
postfix-expression -> template_opt id-expression
```
http://eel.is/c++draft/expr.post#expr.ref-6 determines what is denoted by the member access expression, and that `id-expression` will denote an immediate function whether called through `this->func()` or `func()`.

**Comment by cor3ntin** [comment]
Should be fixed by #63139

**Comment by shafik** [comment]
This currently working in trunk: https://godbolt.org/z/4P8GdPhWM
